apps/myaccount/src/components/consents/consent-edit.tsx

```diff
@@ -50,38 +58,195 @@ export const AppConsentEdit: FunctionComponent<EditConsentProps> = (
+            .filter((piiClaim) => piiClaim.receiptId === editingConsent.consentReceiptID)
+            .value();
+
+        // FIXME: solve in linear time
+        for (const uiRecord of recordOnUserInterface) {
+            for (const { purposeId, piiCategory } of recordOnModelReceipt) {
+                if (purposeId === uiRecord.purposeId) {
+                    for (const piiCategoryKey of piiCategory) {
+                        if (piiCategoryKey.piiCategoryId === uiRecord.piiCategoryId &&
+                            piiCategoryKey.status !== uiRecord.status
+                        ) {
+                            return true;
+                        }
+                    }
+                }
             }
         }
+
+        return false;
     };
 
+    /**
+     * A predicate that checks whether we can show the consents'
+     * purposes. This check is needed because {@link ConsentInterface.consentReceipt}
+     * is optional and will only fetch asynchronously once the user click the detail view.
+     *
+     * {@link ConsentInterface} must: -
+     * 1. not be {@code null | undefined }
+     * 2. contain the {@link ConsentReceiptInterface}
+     * 3. contain a list of {@link ServiceInterface} in {@link ConsentReceiptInterface}
+     * 4. contain least 1 {@link ConsentReceiptInterface.services} entry
+     *
+     * @param {ConsentInterface} editingConsent
+     */
+    const hasConsentDetails = (editingConsent: ConsentInterface): boolean => {
+        return editingConsent &&
+            editingConsent.consentReceipt &&
+            editingConsent.consentReceipt.services &&
+            editingConsent.consentReceipt.services.length > 0;
+    }
+
+    /**
+     * A predicate that checks whether a particular service inside
+     * a {@link ConsentReceiptInterface} has purposes to show.
+     *
+     * @param {ServiceInterface} service
+     */
+    const hasPurposesInService = (service: ServiceInterface): boolean => {
+        return service &&
+            service.purposes &&
+            service.purposes.length > 0;
+    }
+
+    /**
+     * A predicate that checks whether a particular {@link PurposeInterface}
+     * has a list of {@link PIICategory[]} and is not empty.
+     *
+     * @param {PurposeInterface} purpose
+     */
+    const hasPIICategoriesInPurpose = (purpose: PurposeInterface): boolean => {
+        return purpose.piiCategory && purpose.piiCategory.length > 0;
+    }
+
+    /**
+     * JSX Builder method. Invoked within the scope of {@link this}
+     *
+     * @param {PurposeInterface} purpose
+     */
+    const eachPurposeToJSX = (purpose: PurposeInterface) => {
+
+        /**
+         * JSX Builder method. Invoked within the scope of {@link eachPurposeToJSX}
+         * @param {PIICategoryWithStatus} piiCat
+         */
+        const eachPIICategoryItem = (piiCat: PIICategoryWithStatus) => {
+
+            return <List.Item key={ piiCat.piiCategoryId }>
+                <List.Content>
```

Comment:
Fix indentation of the child element so it looks more readable.

Reply:
Applied this improvement in commit 4cb01ba